### ISSUE TITLE
google announces tpu 3.0

### ISSUE THREAD

**Issue by l1t1** [leela-zero/leela-zero#1377]
https://www.reddit.com/r/MachineLearning/comments/8hzsvf/n_google_announces_tpu_30/
can leelazero use their tpu?

**Comment by 2ji3150** [comment]
So you wanna to donate?

**Comment by hydrogenpi** [comment]
So its eight times faster than the gen 2 
But I'm thinking this will only be on the google cloud platform for rent and will be very expensive. 

Not going to be of much help to LZ project unless Google donates lots of hours which isn't likely 
Seeing as they shutdown the colab that lczero was using and/or throttle it to useless

**Comment by jackpal** [comment]
While TPUv3 was announced today, it's not available for use yet.

It's hard for a hobby project like Leela Zero to justify paying to use TPUs as long as there are hobbyists willing to donate their GPU time to the project.  

The earlier TPUv2 is available at $6.50 per TPU per hour. (plus another dollar an hour for CPU, storage, bandwith etc. Plus sales tax I guess.) Each TPUv2 is about as powerful as a V100 (which in turn is about 1.3 x a 1080ti). By comparison the marginal cost of using a hobby PC with a 1080ti is closer to $0.50 an hour, including depreciation, internet service and electricity but not including an administrator or home or office rent. (The hobby PC doesn't have ECC, but I bet leela zero training is resilient to random bit flips that cause bad moves.)

Cloud TPUs make enormously good sense for businesses and researchers, but probably not for computer go hobbyists who already have the sunk cost of a PC and graphics card.

Aside from the higher cost, to use TPUs Leela Zero would have to support TensorFlow in addition to the current OpenCL API.

**Comment by NhanHo** [comment]
@jackpal Every cloud TPU is actually equivalent to 4x V100 (a cloud TPU has 4 chips in it), that makes the maths a bit different still.

For leela zero training process, I've found that V100 is about twice faster than 1080ti. Though I'm not sure how much this will transfer to inference (ie. self play games).

**Comment by jackpal** [comment]
@nhanho Citation? I was going by https://blog.riseml.com/benchmarking-googles-new-tpuv2-121c03b71384

**Comment by NhanHo** [comment]
@jackpal : same author, updated benchmark: https://blog.riseml.com/comparing-google-tpuv2-against-nvidia-v100-on-resnet-50-c2bbb6a51e5e . Specifically you can see this image:  https://cdn-images-1.medium.com/max/1600/0*N6eIjl6FGXzfgpiW.

**Comment by jackpal** [comment]
Yeah, I misunderstood his article, thanks for correcting me.

**Comment by jillybob** [comment]
TPU 3 is 8 times faster than currrent Gen2, so even if prices were $10/hr,  it will still be economical.

**Comment by gcp** [comment]
>For leela zero training process, I've found that V100 is about twice faster than 1080ti.

How did you test? Did you use fp16 mode or something?

**Comment by NhanHo** [comment]
@gcp: the positions/ sec on V100 is twice faster (I was training a 20b network). The speed difference is in `measure_loss` and the 1080ti has higher RAM/ SSD/ CPU speed,  so I am think it is the gpu difference. 

Fp16 was not used.

**Comment by afalturki** [comment]
@gcp Regarding fp16, is it possible to add fp16 mode support as a compilation option such that people who have cards with full fp16 support like a Titan-V or V100 for example can make better use of their cards. Making it a compile option would also insure people enabling it know what they're doing. Although it's unreasonable to expect consumers to have access to a V100, I and other people have successfully rented a V100 for ~4 days from Microsoft Azure using the free $200 they give. I think Google Compute give $300 free credit too. So, having fp16 support for that would be very beneficial.

**Comment by gcp** [comment]
>Fp16 was not used.

It's strange because V100 has very little performance difference from 1080 Ti if fp16 or Tensor Cores are not used, and I think Tensor Cores have to be enabled due to loss of precision.

My experience with training was the opposite.

**Comment by gcp** [comment]
@afalturki We have no fp16 math support so there's nothing to enable. USE_HALF just relates to memory storage and is a minor speedup.

**Comment by jkiliani** [comment]
In discussions about Leela Chess possibly using Tensor cores in the future, I got the impression that TensorRT solves the accuracy problem of using fp16 or even int8 by recalibrating the activation functions (i.e. ReLU) to limit the accuracy loss when doing inference, although that apparently is not possible for training. Is OpenCL likely to support this type of hardware (V100, Or Titan V) in the future, and could it be used for neural net inferencing in Leela Zero?

**Comment by gcp** [comment]
You are really asking 2 completely different questions at once here.

TensorRT is proprietary black magic that does quite a lot besides (vastly) improving the accuracy loss on inference by massaging the network. It's only ever going to work on NVIDIA GPUs, and it's unrelated to OpenCL/CUDA/whatever. You pretty much need a dedicated backend for it.

OpenCL supports fp16 (cl_khr_fp16) and int8 types. So to get fp16 support, someone "just" (cough cough) needs to port the kernel and solve any rounding issues that arise. Same for using int8 inference, although there you will be even more likely to need to replicate NVIDIA's research on dealing with the precision loss. Tensor Cores would need NVIDIA to provide an OpenCL extension for them in their driver. You have to ask them if they have plans for this, but I guess it depends on customer demand.

**Comment by gcp** [comment]
>I meant adding full fp16 math support to LZ and making it a compile option

Well, to answer your question: yes, it is *possible* to do so.

**Comment by Ka-zam** [comment]
>  I meant adding full fp16 math support

This is the thing that is not trivial. You need to write a new fp16 OpenCL kernel instead of the one we have which is fp32.

See: https://github.com/gcp/leela-zero/blob/master/src/OpenCL.cpp

**Comment by gcp** [comment]
The CLBlast kernels actually support fp16, so it's more going to be an issue of making the passed types a bit more generic and making sure they match up. 

The real problems would start if the rounding errors become too great in fp16 and the loss of precision in the Winograd transforms becomes fatal, but you need the above to work before you can know that first.

It's probably quite doable, but you need the overlap between someone who is able to do it, cares enough, and has actual fp16 hardware to test on (hah!).

**Comment by gcp** [comment]
@Ka-zam I'm not entirely sure the rounding problems will be identical if fp16 is emulated. Also, many NVIDIA cards have no support for fp16 whatsoever in OpenCL:

```
 Platform Name                                   NVIDIA CUDA
Number of devices                                 1
  Device Name                                     GeForce GTX 1060 6GB
  Device Vendor                                   NVIDIA Corporation
  Device Vendor ID                                0x10de
  Device Version                                  OpenCL 1.2 CUDA
  Driver Version                                  390.48
  Preferred / native vector sizes
    half                                                 0 / 0        (n/a)
  Half-precision Floating-point support           (n/a)
```

AMD cards report full fp16 capability, even if they don't really have it:

```
  Platform Name                                   Clover
Number of devices                                 1
  Device Name                                     Radeon RX 560 Series (POLARIS11 / DRM 3.23.0 / 4.16.7-1-MANJARO, LLVM 6.0.0)
  Device Vendor                                   AMD
  Device Vendor ID                                0x1002
  Device Version                                  OpenCL 1.1 Mesa 18.0.3
  Driver Version                                  18.0.3
  Preferred / native vector sizes
    half                                                 8 / 8        (cl_khr_fp16)
 Half-precision Floating-point support           (cl_khr_fp16)
    Support is emulated in software               No
```

So, the tuner probably wants to benchmark it.

**Comment by sethtroisi** [comment]
https://github.com/tensorflow/minigo/issues/541
